### PR TITLE
fix: handle DTensor weights in MultiLoRAGroupedExperts for EP

### PR DESCRIPTION
**Pull request by hallerite**
## Summary

When Expert Parallelism is enabled, `distribute_module()` converts all parameters to DTensors sharded across the EP mesh. This PR fixes three issues that prevented EP + LoRA MoE from working end-to-end:

1. **Forward path** (first commit): `MultiLoRAGroupedExperts.forward()` bypassed the `@expert_parallel` decorator, passing DTensors directly to `torch._grouped_mm`. Fix: mirror the decorator logic — convert DTensors to local shards, permute/pad tokens, and unpermute output.

2. **Weight broadcast** (`state_dict_for_adapter`): Per-expert indexing on EP-sharded DTensor LoRA weights triggered a `_redistribute` assertion. Fix: call `full_tensor()` to gather sharded weights before indexing.

3. **LoRA config** (`save_lora_config`): After EP/FSDP wrapping, `base_layer` params can have `requires_grad=True`, incorrectly adding them to `modules_to_save` — which vLLM rejects. Fix: exclude `base_layer` from the detection.

## Test plan

- [x] Qwen3-30B-A3B-Instruct EP=4, LoRA rank 16, batch_size=128, 30 steps — completed successfully
- [x] Same config with batch_size=256 — completed successfully
- [x] Both runs on single H200 node (4 train + 4 infer GPUs, TP=4 inference)
- [x] Rewards improving across steps (0.22 → 0.55), weight broadcast working, vLLM LoRA loading working
- [x] W&B: https://wandb.ai/primeintellect/test-moe-lora-ep4

---

> [!NOTE]
> **Medium Risk**
> Touches expert-parallel MoE forward and LoRA weight export paths; incorrect permutation/padding or DTensor gathering could cause wrong outputs or memory/perf regressions under EP.
> 
> **Overview**
> Fixes EP + LoRA MoE interoperability when parameters are converted to sharded `DTensor`s.
> 
> `MultiLoRAGroupedExperts` now gathers sharded LoRA weights when exporting per-expert adapter state, and its `forward()` mirrors EP token permutation/padding by converting `DTensor` weights to local shards before `torch._grouped_mm`, then unpermuting outputs back to dispatched token order.
> 
> During model setup, LoRA freezing is re-applied after `apply_ep()` to counter EP replacing parameters with `requires_grad=True` `DTensor`s, ensuring only LoRA (and configured `modules_to_save`) remain trainable.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6fec5a07df4b269666624925d402af3eb2a80f95. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/models/layers/lora/multi_moe.py [hunk 1]
```diff
@@ -66,6 +68,76 @@ def _run_lora_for_loop(
...
+        h_lora = F.dropout(h, p=dropout_p, training=training) if dropout_p > 0.0 else h
+        h2_base = torch.matmul(h, base_w2[expert_idx].transpose(-2, -1))
+        w2_lora_tmp = torch.matmul(h_lora, w2_lora_a[expert_idx].transpose(-2, -1))
+        w2_lora_out = torch.matmul(w2_lora_tmp, w2_lora_b[expert_idx].transpose(-2, -1))
+        out = h2_base + scaling * w2_lora_out
+
+        out_splits.append(out)
+        start = end
+
+    return torch.cat(out_splits, dim=0)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lora/multi_moe.py]: yea we dont care to support EP for the for loop path. lets just raise at apply ep time if someone tries to do for loop with EP

File: src/prime_rl/trainer/lora.py [hunk 2]
```diff
@@ -260,7 +260,7 @@ def save_lora_config(model: nn.Module, save_path, rank: int, alpha: float, dropo
             target_modules.add(module_suffix)
 
     for name, param in model.named_parameters():
-        if param.requires_grad and "lora_A" not in name and "lora_B" not in name:
```
Comment (Jackmin801) [inline src/prime_rl/trainer/lora.py:263]: oh was this wrong before?

**Review by Jackmin801 — APPROVED** [review]
Nice! lfgtm!

File: src/prime_rl/trainer/models/layers/lora/multi_moe.py [hunk 3]
```diff
@@ -363,6 +406,19 @@ def forward(self, x: torch.Tensor, num_tokens_per_expert: torch.Tensor) -> torch
...
             )
 
+        # EP handling: unpermute output back to dispatched token order
+        if permuted_indices is not None:
+            # For-loop may produce fewer rows than permuted input (no padding rows),
+            # pad to match before unpermuting
+            if out.shape[0] < len(permuted_indices):
+                num_padding = len(permuted_indices) - out.shape[0]
+                out = torch.vstack((out, out.new_zeros((num_padding, out.shape[-1]))))
+            out_unpermuted = out.new_empty(input_shape)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lora/multi_moe.py]: hrmm this might be valid
